Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -915,6 +962,122 @@ lemma nonsingular [Nontrivial R] {x y : R} (h : E.equation x y) : E.nonsingular
   E.nonsingular_of_Δ_ne_zero h <| E.coe_Δ' ▸ E.Δ'.ne_zero
 #align elliptic_curve.nonsingular EllipticCurve.nonsingular
 
+section ModelsWithJ
+
+variable (R)
+
+/-- When $3$ is invertible, $Y^2 + Y = X^3$ is an elliptic curve.
+It is of $j$-invariant $0$ (see `EllipticCurve.ofJ0_j`). -/
+def ofJ0 [Invertible (3 : R)] : EllipticCurve R :=
+  have := invertibleNeg (3 ^ 3 : R)
+  ⟨WeierstrassCurve.ofJ0 R, unitOfInvertible (-3 ^ 3 : R),
+    by rw [unitOfInvertible_val, WeierstrassCurve.ofJ0_Δ R]; norm_num1⟩
+
+lemma ofJ0_j [Invertible (3 : R)] : (ofJ0 R).j = 0 := by
+  simp only [j, ofJ0, WeierstrassCurve.ofJ0_c₄]
+  ring1
+
+/-- When $2$ is invertible, $Y^2 = X^3 + X$ is an elliptic curve.
+It is of $j$-invariant $1728$ (see `EllipticCurve.ofJ1728_j`). -/
+def ofJ1728 [Invertible (2 : R)] : EllipticCurve R :=
+  have := invertibleNeg (2 ^ 6 : R)
+  ⟨WeierstrassCurve.ofJ1728 R, unitOfInvertible (-2 ^ 6 : R),
+    by rw [unitOfInvertible_val, WeierstrassCurve.ofJ1728_Δ R]; norm_num1⟩
+
+lemma ofJ1728_j [Invertible (2 : R)] : (ofJ1728 R).j = 1728 := by
+  field_simp [j, ofJ1728, @unitOfInvertible_val _ _ _ <| invertibleNeg _,
+    WeierstrassCurve.ofJ1728_c₄]
+  norm_num1
+
+variable {R}
+
+/-- When $j$ and $j - 1728$ are both invertible,
+$Y^2 + (j - 1728)XY = X^3 - 36(j - 1728)^3X - (j - 1728)^5$ is an elliptic curve.
+It is of $j$-invariant $j$ (see `EllipticCurve.ofJ'_j`). -/
+def ofJ' (j : R) [Invertible j] [Invertible (j - 1728)] : EllipticCurve R :=
+  have := invertibleMul (j ^ 2) ((j - 1728) ^ 9)
+  ⟨WeierstrassCurve.ofJ j, unitOfInvertible <| j ^ 2 * (j - 1728) ^ 9,
+    (WeierstrassCurve.ofJ_Δ j).symm⟩
+
+lemma ofJ'_j (j : R) [Invertible j] [Invertible (j - 1728)] : (ofJ' j).j = j := by
+  field_simp [EllipticCurve.j, ofJ', @unitOfInvertible_val _ _ _ <| invertibleMul _ _,
+    WeierstrassCurve.ofJ_c₄]
+  ring1
+
+variable {F : Type u} [Field F] (j : F)
+
+private lemma two_or_three_ne_zero : (2 : F) ≠ 0 ∨ (3 : F) ≠ 0 :=
+  ne_zero_or_ne_zero_of_nat_coprime (show Nat.coprime 2 3 by norm_num1)
+
+variable [DecidableEq F]
+
+/-- For any element $j$ of a field $F$, there exists an elliptic curve over $F$
+with $j$-invariant equal to $j$ (see `EllipticCurve.ofJ_j`).
+Its coefficients are given explicitly (see `EllipticCurve.ofJ0`, `EllipticCurve.ofJ1728`
+and `EllipticCurve.ofJ'`). -/
+def ofJ : EllipticCurve F :=
+  if h0 : j = 0 then
+    if h3 : (3 : F) = 0 then @ofJ1728 _ _ <| invertibleOfNonzero <|
+      two_or_three_ne_zero.neg_resolve_right h3
+    else @ofJ0 _ _ <| invertibleOfNonzero h3
+  else if h1728 : j = 1728 then
+    @ofJ1728 _ _ <| invertibleOfNonzero fun h => h0 <|
+    by rw [h1728, show (1728 : F) = 2 * 864 by norm_num1, h, zero_mul]
+  else @ofJ' _ _ j (invertibleOfNonzero h0) (invertibleOfNonzero <| sub_ne_zero_of_ne h1728)
+
+lemma ofJ_0_of_three_ne_zero (h3 : (3 : F) ≠ 0) : ofJ 0 = @ofJ0 _ _ (invertibleOfNonzero h3) := by
```

Comment:
What about using [NeZero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/NeZero.html#NeZero) instead of `h3`? Similarly elsewhere.